pkg/controller/release/strategy_executor_test.go

```diff
@@ -0,0 +1,339 @@
+package release
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
+	shippertesting "github.com/bookingcom/shipper/pkg/testing"
+	"github.com/bookingcom/shipper/pkg/util/conditions"
+)
+
+type fakePipeline struct {
+	steps     []string
+	processed string
+}
+
+func (p *fakePipeline) increase(ctx *context, firstRelInfo, secondRelInfo *releaseInfo) {
+	p.Enqueue(genCapacityEnforcer(ctx, firstRelInfo, secondRelInfo))
+	p.Enqueue(genTrafficEnforcer(ctx, firstRelInfo, secondRelInfo))
+}
+
+func (p *fakePipeline) decrease(ctx *context, firstRelInfo, secondRelInfo *releaseInfo) {
+	p.Enqueue(genTrafficEnforcer(ctx, firstRelInfo, secondRelInfo))
+	p.Enqueue(genCapacityEnforcer(ctx, firstRelInfo, secondRelInfo))
+}
+
+func (p *fakePipeline) Enqueue(step PipelineStep) {
+	functionName := getFunctionName(step)
+	p.steps = append(p.steps, functionName)
+}
+
+func getFunctionName(i interface{}) string {
+	fullName := runtime.FuncForPC(reflect.ValueOf(i).Pointer()).Name()
+	relativeName := filepath.Base(fullName)
+	split := strings.Split(relativeName, ".")
+	return split[1]
+}
+
+func (p *fakePipeline) Process(strategyStep shipper.RolloutStrategyStep, cond conditions.StrategyConditionsMap) (bool, []StrategyPatch, []ReleaseStrategyStateTransition) {
+	p.processed = strings.Join(p.steps, "\n")
+	return false, nil, nil
+}
+
+const testPrevRelease = "test-prev-release"
+const testCurrRelease = "test-curr-release"
+const testSuccRelease = "test-succ-release"
+const testNamespace = "test-namespace"
+const succRelUid = "random-succ-uid"
+const currRelUid = "random-curr-uid"
+const prevRelUid = "random-prev-uid"
+
+func TestExecute(t *testing.T) {
+	var tests = []struct {
+		name                string
+		prev, curr, succ    *releaseInfo
+		isSteppingBackwards bool
+		expectedOut         string
```

Comment:
Could we make this an `[]string` instead? It will be more readable in test cases IMO.